ext/opentelemetry-ext-opentracing-shim/src/opentelemetry/ext/opentracingshim/__init__.py

```diff
@@ -0,0 +1,232 @@
+# Copyright 2019, OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from contextlib import contextmanager
+
+import opentracing
+
+from opentelemetry.ext.opentracingshim import util
+from opentelemetry.trace import Event
+
+
+def create_tracer(tracer, scope_manager=None):
+    return TracerWrapper(tracer, scope_manager)
+
+
+class SpanContextWrapper(opentracing.SpanContext):
+    def __init__(self, otel_context):
+        self._otel_context = otel_context
+
+    def unwrap(self):
+        """Returns the wrapped OpenTelemetry `SpanContext` object."""
+
+        return self._otel_context
+
+    @property
+    def baggage(self):
+        raise NotImplementedError
+        # TODO: Implement.
+
+
+class SpanWrapper(opentracing.Span):
+    def __init__(self, tracer, context, span):
+        super().__init__(tracer, context)
+        self._otel_span = span
+
+    def unwrap(self):
+        """Returns the wrapped OpenTelemetry `Span` object."""
+
+        return self._otel_span
+
+    def set_operation_name(self, operation_name):
+        self._otel_span.update_name(operation_name)
+
+        # Return self for call chaining.
+        return self
+
+    def finish(self, finish_time=None):
+        end_time = finish_time
+        if end_time is not None:
+            end_time = util.time_seconds_to_ns(finish_time)
+        self._otel_span.end(end_time=end_time)
+
+    def set_tag(self, key, value):
+        self._otel_span.set_attribute(key, value)
+
+        # Return self for call chaining.
+        return self
+
+    def log_kv(self, key_values, timestamp=None):
+        if timestamp is not None:
+            event_timestamp = util.time_seconds_to_ns(timestamp)
+        else:
+            event_timestamp = util.time_ns()
+
+        event_name = util.event_name_from_kv(key_values)
+        event = Event(event_name, event_timestamp, key_values)
+        self._otel_span.add_lazy_event(event)
+
+        # Return self for call chaining.
+        return self
+
+    def set_baggage_item(self, key, value):
+        raise NotImplementedError
+        # TODO: Implement.
+
+    def get_baggage_item(self, key):
+        raise NotImplementedError
+        # TODO: Implement.
+
+    # TODO: Verify calls to deprecated methods `log_event()` and `log()` on
+    # base class work properly (it's probably fine because both methods call
+    # `log_kv()`).
+
+
+class ScopeWrapper(opentracing.Scope):
```

Comment:
Maybe too pedantic a point, but the classes above look like they follow the naming scheme:

```python
class XWrapper(opentracing.Y)   # wraps opentelemetry.trace.X`
```

and this one follows:

```python
class XWrapper(opentracing.X)   # wraps opentelemetry.trace.Y`
```

It makes sense in context for e.g. `SpanWrapper` to take an `opentelemetry.trace.Span` argument, but there's nothing for `ScopeWrapper` to "wrap" here since `opentelemetry` doesn't have a scope class. Which makes the other names confusing in retrospect.

Why not just call this `Scope`?

Reply:
I thought the `XWrapper` is trying to wrap `opentracing.X` consistently across this PR, did I miss something?

Reply:
As mentioned somewhere else, we need to handle the case where you don't actually finish a `Span` as part of closing the `Scope` ;)

Reply:
What I had in mind was: `XWrapper` is a class which inherits from `opentracing.X` and hides an `opentelemetry.trace.X` in its implementation.

@c24t is right - `ScopeWrapper` doesn't really wrap an OpenTelemetry scope (because there is no such a thing). Same for `ScopeManagerWrapper` - it doesn't wrap an OpenTelemetry `ScopeManager`.

>Why not just call this `Scope`?

I've actually chosen this name to avoid having a class named `Scope` in both the shim and OpenTracing. Things are already confusing enough as it is since we have a lot of imperfect overlap between OT and OTel.

That said, I am fine with renaming both `ScopeWrapper` and `ScopeManagerWrapper` if this makes more sense.

Reply:
> I thought the `XWrapper` is trying to wrap `opentracing.X` consistently across this PR

I took "wraps X" to mean "forwards calls to X", so these classes *implement* `opentracing.X` but (where applicable) *wrap* `opentelemetry.X`.

> Things are already confusing enough as it is since we have a lot of imperfect overlap between OT and OTel.

Yeah, that makes sense.

> That said, I am fine with renaming both `ScopeWrapper` and `ScopeManagerWrapper` if this makes more sense.

Feel free to ignore this comment if it's clear to you and other reviewers. If you do change the names, replacing "Wrapper" with "Shim" where the class doesn't wrap an OpenTelemetry object might make this clearer.

Reply:
Maybe we should rename everything to "Shim" and when the class wraps an OTel class, we note it in the docs.

Reply:
>Maybe we should rename everything to "Shim"

Done.